File: enterprise-platform/enterprise-platform/svc-product-catalog/src/metrics.js
// Prometheus metrics. These feed the SLOs defined in platform-ops/slo/.
const client = require('prom-client');

const register = new client.Registry();
client.collectDefaultMetrics({ register });

const httpRequestDuration = new client.Histogram({
  name: 'http_request_duration_seconds',
  help: 'HTTP request latency in seconds',
  labelNames: ['method', 'route', 'status_code'],
  // Buckets chosen around our 300ms latency SLO
  buckets: [0.005, 0.01, 0.025, 0.05, 0.1, 0.2, 0.3, 0.5, 1, 2, 5],
});

const httpRequestsTotal = new client.Counter({
  name: 'http_requests_total',
  help: 'Total HTTP requests',
  labelNames: ['method', 'route', 'status_code'],
});

const dependencyRequestsTotal = new client.Counter({
  name: 'dependency_requests_total',
  help: 'Calls made to downstream services',
  labelNames: ['dependency', 'status'],
});

register.registerMetric(httpRequestDuration);
register.registerMetric(httpRequestsTotal);
register.registerMetric(dependencyRequestsTotal);

// Express middleware
function metricsMiddleware(req, res, next) {
  const end = httpRequestDuration.startTimer();
  res.on('finish', () => {
    const route = req.route ? req.route.path : req.path;
    const labels = {
      method: req.method,
      route,
      status_code: res.statusCode,
    };
    end(labels);
    httpRequestsTotal.inc(labels);
  });
  next();
}

module.exports = {
  register,
  metricsMiddleware,
  dependencyRequestsTotal,
};

File: enterprise-platform/enterprise-platform/svc-recommendation/src/metrics.py
"""Prometheus metrics. Names match the Node services deliberately so one
Grafana dashboard and one set of SLO rules cover the whole platform."""
import time

from prometheus_client import Counter, Histogram
from starlette.middleware.base import BaseHTTPMiddleware

HTTP_REQUEST_DURATION = Histogram(
    "http_request_duration_seconds",
    "HTTP request latency in seconds",
    ["method", "route", "status_code"],
    buckets=(0.005, 0.01, 0.025, 0.05, 0.1, 0.2, 0.3, 0.5, 1, 2, 5),
)

HTTP_REQUESTS_TOTAL = Counter(
    "http_requests_total",
    "Total HTTP requests",
    ["method", "route", "status_code"],
)

DEPENDENCY_REQUESTS_TOTAL = Counter(
    "dependency_requests_total",
    "Calls made to downstream services",
    ["dependency", "status"],
)


class MetricsMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request, call_next):
        start = time.perf_counter()
        response = await call_next(request)
        elapsed = time.perf_counter() - start

        # Use the route template, not the raw path, or you get unbounded
        # cardinality (one time series per order id - a classic outage).
        route = request.scope.get("route")
        route_path = route.path if route else request.url.path

        labels = {
            "method": request.method,
            "route": route_path,
            "status_code": str(response.status_code),
        }
        HTTP_REQUEST_DURATION.labels(**labels).observe(elapsed)
        HTTP_REQUESTS_TOTAL.labels(**labels).inc()
        return response
